### PR TITLE
ImageVAETrainerにlogger追加

### PR DESCRIPTION
**Pull request by myxyy**
## 概要

https://github.com/MLShukai/ami/issues/99
Trainerのlogger実装の方針の確認です

* updateタイミングは適切か
* ログフォルダ名は適切か

あたりが気になっています

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

### CODE REVIEW

File: configs/trainers/image_vae.yaml [hunk 1]
```diff
@@ -9,5 +9,9 @@
     _partial_: true
     lr: 0.001
   device: ${devices.0}
+  logger:
+    _target_: ami.tensorboard_loggers.StepIntervalLogger
+    log_dir: ${paths.log_dir}/tensorboard
```
Comment (Geson-anko) [inline configs/trainers/image_vae.yaml:14]: `ami/configs/paths/default.yaml`に`tensorboard_dir: ${paths.log_dir}/tensorboard`を追加してください。そして以下のようにすると良いと思います。Trainerごとに個別のTensorBoardディレクトリを用意する理由は、再開した際にその個別フォルダに新規のログファイルが生成され、引継が行われるというTensorBoardの仕様のためです。

```suggestion
    log_dir: ${paths.tensorboard_dir}/image_vae_trainer
```

File: ami/trainers/image_vae_trainer.py [hunk 2]
```diff
@@ -81,9 +84,11 @@ def train(self) -> None:
                     dist_batch, Normal(torch.zeros_like(dist_batch.mean), torch.ones_like(dist_batch.stddev))
                 )
                 loss = rec_loss + self.kl_coef * kl_loss.mean()
+                self.logger.log("image_vae_loss", loss)
```
Comment (Geson-anko) [inline ami/trainers/image_vae_trainer.py:87]: このようにすると、`image_vae`というタブの中に`loss`がまとめられて見やすくなると思います。
```suggestion
                self.logger.log("image_vae/loss", loss)
```

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@myxyy 実装ありがとうございます。loggerの`update`メソッドはステップごとに呼んでください。もし`update`というメソッド名が不適切だと思う場合は、`step`など、適切なものに置き換えてしまって構いません。

あと`tensorboard`の`global_step`も状態なので、それを`save_state`/`load_state`できるようにしておく必要があるかもしれません。

File: ami/trainers/image_vae_trainer.py [hunk 3]
```diff
@@ -81,9 +84,11 @@ def train(self) -> None:
...
                 loss.backward()
                 optimizer.step()
 
+        self.logger.update()
```
[the remark this answers is not in the document]
Comment (myxyy) [reply ami/trainers/image_vae_trainer.py]: https://github.com/MLShukai/ami/pull/158/commits/53e29e7f6c301c8997e92807f089f0ad0a62f148
epochのループ内ではなくbatchのループ内でよいでしょうか？修正しました

File: configs/trainers/image_vae.yaml (hunk 1 above)
Comment (myxyy) [reply configs/trainers/image_vae.yaml]: https://github.com/MLShukai/ami/pull/158/commits/da4036c8674ea4183902e5d9503779873220a489
修正しました

File: ami/trainers/image_vae_trainer.py (hunk 2 above)
Comment (myxyy) [reply ami/trainers/image_vae_trainer.py]: https://github.com/MLShukai/ami/pull/158/commits/f8c22ce7210ac142cc6a715c08b8710e7e9f63c4
修正しました

**Comment by myxyy** [comment]
https://github.com/MLShukai/ami/pull/158/commits/5b72324fd8f717191ee1a6e7620085d8aa4d7ab1
loggerのglobal_stepを保存するようにしました

File: ami/trainers/image_vae_trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Geson-anko) [reply ami/trainers/image_vae_trainer.py]: @myxyy `logger.update()`は`optimizer.step()`に対応するものだと考えてください。

**Comment by Geson-anko** [comment]
@myxyy テストが通らない原因なのですが、GitHubのテスト実行はどうやらマージ後のリポジトリの状態を再現して実行するらしく、mainブランチへのマージ後、 `configs/trainers/image_vae_forward_dynamics_ppo.yaml`にTensorboardLoggerが追加されていないためにエラーが出てしまったようです。

**Comment by Geson-anko** [comment]
それぞれのTrainerのベースの設定ファイルが共有されないと困るので、ちょっとそこ修正します。

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@myxyy 修正ありがとうございます。logの出力ディレクトリだけ修正した後にマージしましょう。

**Review by Geson-anko — APPROVED** [review]
手隙だったので`output_dir`への修正も行いました。https://github.com/MLShukai/ami/pull/158/commits/054974e34bae8fb3765cd6de7b9ac167426c2977

ただ、メイン実行したランタイムでしか`hydra:runtime.output_dir`が使えないようで、テストでそこが詰まるので修正しました。
